### DISCUSSION TITLE
Is this compatible with CryptoKit?

### DISCUSSION THREAD

**Discussion by thejeraldo** [krzyzanowskim/CryptoSwift#999] [Q&A]
We are trying to encrypt and decrypt a string in iOS and Android.
Both iOS and Android can encrypt and decrypt the same data from each other, but I'm curious as to why **CryptoKit** AES GCM is different and we just couldn't match between the same platform? In **CryptoSwift** iOS to Android and Android to iOS AES CBC works.

Thanks.

**Comment by nathanfallet** [comment]
Can you provide some code to reproduce?

**Comment by vinicemanuel** [comment]
hello, I have same problem 
here some code do reproduce

@Suite("CertificateEncryptor tests")
struct SwiftLabiOSTests {
    
    private let key = "101|102|52|102|102|48|102|49|102|99|51|49|49|97|97|48|102|50|50|98|98|54|52|53|57|53|49|56|99|102|57"
    private let iv = "73|107|73|90|106|89|73|97|103|106|78|51|118|81|81|85"
    
    @Test("compare encryptToSplitBytes")
    func test() async throws {
        let certificateEncryptor = CertificateEncryptor(
            key: key,
            initializationVector: iv
        )
        
        let pinCrypto = certificateEncryptor.encryptToSplitBytes("1111")
        
        let certificateEncryptorNew = CertificateEncryptorNew(
            key: key,
            initializationVector: iv
        )
        
        let pinCryptoNew = certificateEncryptorNew.encryptToSplitBytes("1111")
        
        #expect(pinCrypto == pinCryptoNew)
    }
}


import Foundation
import CryptoKit
 
final class CertificateEncryptorNew {
    let key: String
    let initializationVector: String
    
    //MARK: - Init
 
    init(key: String, initializationVector: String) {
        self.key = key
        self.initializationVector = initializationVector
    }
    
    //MARK: - Internal
    func encryptToSplitBytes(_ content: String) -> String {
        let encrypted = encrypt(stringToEncrypt: content)
        let splitBytes = getSplitBytesUnsigned(encrypted)
        return splitBytes
    }
    
    //MARK: - Static
    
    //MARK: - Private
    
    private func getSplitBytesUnsigned(_ encrypted: String) -> String {
        var result = ""
        for char in encrypted.utf8 {
            result += String(char) + "|"
        }
        result = String(result[..<result.index(result.endIndex, offsetBy: -1)])
        return result
    }
    
    private func encrypt(stringToEncrypt: String) -> String {
        do {
            var keyBytes: [UInt8] = Array(key.utf8)
            var ivBytes: [UInt8] = Array(initializationVector.utf8)
            
            if keyBytes.count > 16 {
                keyBytes = Array(key.utf8.prefix(16))
            }
            
            if ivBytes.count > 12 {
                ivBytes = Array(initializationVector.utf8.prefix(12))
            }
            
            let symmetricKey = SymmetricKey(data: Data(keyBytes))
            let nonce = try AES.GCM.Nonce(data: Data(ivBytes))
            
            let plaintext = Data(stringToEncrypt.utf8)
            let sealedBox = try AES.GCM.seal(plaintext, using: symmetricKey, nonce: nonce)
            
            return sealedBox.combined?.base64EncodedString() ?? ""
        } catch {
            print(error.localizedDescription)
        }
        
        return ""
    }
}


import Foundation
import CryptoSwift
 
final class CertificateEncryptor {
    let key: String
    let initializationVector: String
 
    init(key: String, initializationVector: String) {
        self.key = key
        self.initializationVector = initializationVector
    }
 
 
    func encryptToSplitBytes(_ content: String) -> String {
        let encrypted = encrypt(stringToEncrypt: content)
        let spliteBytes = getSplitBytesUnsigned(encrypted)
        return spliteBytes
    }
 
    private func getSplitBytesUnsigned(_ encrypted: String) -> String {
        var result = ""
        for char in encrypted.utf8 {
            result += String(char) + "|"
        }
        result = String(result[..<result.index(result.endIndex, offsetBy: -1)])
        return result
    }
 
    private func encrypt(stringToEncrypt: String) -> String {
        var result = ""
        do {
            var keyBytes: [UInt8] = Array(key.utf8)
            var ivBytes: [UInt8] = Array(initializationVector.utf8)
 
            if keyBytes.count > 32 {
                keyBytes = Array(key.utf8.prefix(32))
            }
 
            if ivBytes.count > 16 {
                ivBytes = Array(initializationVector.utf8.prefix(16))
            }
 
            let gcm = GCM(iv: ivBytes, tagLength: 128, mode: .combined)
            let aes = try AES(key: keyBytes, blockMode: gcm, padding: .noPadding)
 
            let encryptedBytes = try aes.encrypt(Array(stringToEncrypt.utf8))
            result = Data(encryptedBytes).base64EncodedString()
        } catch {
            print(error.localizedDescription)
        }
        return result
    }
}
